clang/lib/CIR/CodeGen/CIRGenExprConst.cpp

```diff
@@ -1717,17 +1717,29 @@ mlir::Attribute ConstantEmitter::tryEmitPrivateForVarInit(const VarDecl &D) {
   // initialization of memory to all NULLs.
   if (!D.hasLocalStorage()) {
     QualType Ty = CGM.getASTContext().getBaseElementType(D.getType());
-    if (Ty->isRecordType())
+    if (Ty->isRecordType()) {
       if (const CXXConstructExpr *E =
               dyn_cast_or_null<CXXConstructExpr>(D.getInit())) {
         const CXXConstructorDecl *CD = E->getConstructor();
         // FIXME: we should probably model this more closely to C++ than
         // just emitting a global with zero init (mimic what we do for trivial
         // assignments and whatnots). Since this is for globals shouldn't
         // be a problem for the near future.
-        if (CD->isTrivial() && CD->isDefaultConstructor())
+        if (CD->isTrivial() && CD->isDefaultConstructor()) {
+          const auto *cxxrd =
+              cast<CXXRecordDecl>(Ty->getAs<RecordType>()->getDecl());
+          // Some cases, such as member pointer members, can't be zero
```

Comment:
Can you clarify that this is an ABI thing? Data member pointers *are* [zero initialized in the standard sense](http://eel.is/c++draft/dcl.init.general#footnote-78). It's just that most popular ABIs including [Itanium](https://itanium-cxx-abi.github.io/cxx-abi/abi.html#data-member-pointers) have a non-zero bit representation of a null data member pointer